Add support for Gen3/Gen4 loaders for Xen DomD

firscity opened the pull request.

firscity committed:
domd: xen: add new board - DomD Xen guest for Renesas Gen3

Add special board for starting Xen driver domain on R-Car Gen3
setups. Separate board is needed to enable both Xen support and
Renesas device drivers in U-boot.

rcar3_xen board allows to boot domain images from ethernet or MMC
devices.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Acked-by: Sergiy Kibrik <Sergiy_Kibrik@epam.com>

sa-kib committed:
ufs: port linux driver for rcar ufshcd

Add driver for initial support of Renesas UFS host controller.
Sources for port were taken from v5.10.41/rcar-5.1.7.rc6 Renesas
Linux BSP branch.

Signed-off-by: Sergiy Kibrik <Sergiy_Kibrik@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

sa-kib committed:
ufs: flush/invalidate command buffer

Cache maintenance required on platforms other than TI (which maitains cache
coherency out of the box using L3 interconnect).

Author: Sergiy Kibrik <Sergiy_Kibrik@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity committed:
rcar3_xen: add AoS specific bootargs and bootcommand

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Acked-by: Sergiy Kibrik <Sergiy_Kibrik@epam.com>

firscity committed:
boards: renesas: xen: add board for Xen DomD on R-Car Gen4

Add separate board for running U-boot as intermediate bootloader
for Xen driver domain on Renesas Gen4 Spider board. It is needed
to use Renesas drivers for peripherals, that is shared with DomD.
As these driver is depandand on ARCH_RMOBILE and other Renesas
Kconfig, it was needed to merge Xen and Renesas configs.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Acked-by: Sergiy Kibrik <Sergiy_Kibrik@epam.com>

firscity committed:
rcar4_xen: add AoS specific bootargs and bootcommand

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Acked-by:  Sergiy Kibrik <Sergiy_Kibrik@epam.com>

sa-kib commented:
This fixup should also be applied:
https://github.com/sa-kib/u-boot-rcar/commit/976ff196b33e6d866dcccc73ab3c0ca40841a2c9

sa-kib commented on drivers/ufs/ufs.c:
```diff
@@ -1442,8 +1456,42 @@ static void prepare_prdt_table(struct ufs_hba *hba, struct scsi_cmd *pccb)
+		len_to_copy = min_t(int, RESPONSE_UPIU_SENSE_DATA_LENGTH, len);
+
+		memcpy(sense_buffer, ucd_rsp_ptr->sr.sense_data, len_to_copy);
+	}
+}
```
Code of `ufshcd_get_rsp_upiu_data_seg_len()` & `ufshcd_copy_sense_data()` routines not needed, remove it from this patch

sa-kib commented on board/renesas/rcar3-xen/rcar3-xen.c:
```diff
@@ -0,0 +1,215 @@
+	xen_mem_map[i].size = 0x40000000UL;
+	xen_mem_map[i].attrs = PTE_BLOCK_MEMTYPE(MT_DEVICE_NGNRNE) |
+			       PTE_BLOCK_NON_SHARE | PTE_BLOCK_PXN |
+			       PTE_BLOCK_UXN;
```
nice to have a comment here describing this range

firscity committed:
WA: renesas: cpg: xen: do not disable clocks before Linux

Baremetal U-boot tries to disable all clocks from CPG clock table
before jumping to Linux. Since U-boot may be used as intermediate
bootloader for Xen driver domain on Renesas boards, we should not
disable all clocks (only for peripherals, that is assigned to driver
domain).

Add workaround, that skips clock disabling for Xen builds.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Acked-by:  Sergiy Kibrik <Sergiy_Kibrik@epam.com>

firscity commented:
@sa-kib all comments are fixed now, thank you for review

LKomaryanskiy commented:
@firscity good job!
`Acked-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>`

sa-kib commented:
LGTM
